src/mpid/ch4/netmod/ofi/ofi_types.h

```diff
@@ -397,6 +350,23 @@ typedef struct {
     char kvsname[MPIDI_KVSAPPSTRLEN];
     char pname[MPI_MAX_PROCESSOR_NAME];
     int port_name_tag_mask[MPIR_MAX_CONTEXT_MASK];
+
+    /* Capability settings */
+#ifdef MPIDI_OFI_ENABLE_RUNTIME_CHECKS
```

Comment:
I'm not sure that there will always be one that matters. The list of providers that's spit back out seems like a bit of a mystery to me, but I'm not sure we can count on there always being one that we want in the default list fi we don't provide any hints. Maybe we can ask someone from the OFI group more about this.

I agree that we should use `MPIR_CVAR_OFI_PROVIDER`. If we want to do that, we'll need to double declare all of the macros in `ofi_capability_sets.h` so we can refer to them independent of the `MPIDI_OFI_ENABLE_something` macros. We'd use these macros to initialize the `MPIDI_Global.settings` struct and then overwrite with any selections from the user via `MPIR_CVAR_CH4_OFI_ENABLE_something` variables (if the user changes the value away from the default value).

I don't think your example cases are all valid, specifically 2 and 3. If you pass `psm2` into the `fi_info` call as a hint, you are passing in a hints object which means you're going to have passed in values for all of the hints (even if they're garbage). I think that means you're not going to get options for your psm2 provider anymore, but whatever is closest to your hints.

I don't think my changes are as complicated as they may sound. Let me finish coding them up and let you take a look. If you don't like them, we can make changes from there.